src/Oscoin/Protocol/Sync.hs

```diff
@@ -0,0 +1,393 @@
+                  , Show (PublicKey c)
+                  ) => Show (SyncEvent c tx s)
+
+data SyncError =
+      RequestTimeout ProtocolRequest Addr Timeout
+      -- ^ The given peer exceeded the request timeout when serving this
+      -- 'SyncProtocolRequest'.
+    | NoActivePeers
+      -- ^ There are not active peers to talk to.
+    | NoRemoteTipFound
+      -- ^ It was not possible to fetch a valid tip from our peers.
+    | AllPeersTimeoutError
+      -- ^ All the queried peers didn't respond in the allocated timeout.
+    deriving (Show, Eq)
+
+instance Exception SyncError
+
+data Timeout =
+    MaxTimeoutExceeded (Expected Duration)
+    -- ^ The operation exceeded the expected timeout, in nanoseconds.
+    deriving (Show, Eq)
+
+-- | A closed (inclusive) range [lo,hi].
+data Range = Range
+    { start :: Height
+    , end   :: Height
+    } deriving Show
+
+data ProtocolRequest =
+      GetTip
+    -- ^ Get the tip of the best chain.
+    | GetBlocks
+    -- ^ Get some blocks.
+    | GetBlockHeaders
+    -- ^ Get some block headers.
+    deriving (Show, Eq)
+
+-- | The singleton type for 'ProtocolRequest', capable of carrying some
+-- witness (in this case, 'ProtocolRequest').
+data SProtocolRequest r where
+  SGetTip :: SProtocolRequest 'GetTip
+  SGetBlocks :: SProtocolRequest 'GetBlocks
+  SGetBlockHeaders :: SProtocolRequest 'GetBlockHeaders
+
+-- | A closed type family mapping each 'ProtocolRequest' to its list of
+-- arguments, so that we can use the 'ProtocolRequest' (where all constructors
+-- have kind /*/) as a constraint.
+type family ProtocolRequestArgs (c :: ProtocolRequest) :: * where
+    ProtocolRequestArgs 'GetTip          = ()
+    ProtocolRequestArgs 'GetBlocks       = Range
+    ProtocolRequestArgs 'GetBlockHeaders = Range
+
+-- | A protocol response, indexed by the corresponding 'ProtocolRequest, so
+-- that matching the wrong type will result in a type error.
+type family ProtocolResponse c tx s (r :: ProtocolRequest) :: *
+
+type instance ProtocolResponse c (Tx c) PoW 'GetTip =
+        Block c (Tx c) (Sealed c PoW)
+type instance ProtocolResponse c (Tx c) PoW 'GetBlocks =
+        OldestFirst [] (Block c (Tx c) (Sealed c PoW))
+type instance ProtocolResponse c tx PoW 'GetBlockHeaders =
+        OldestFirst [] (BlockHeader c (Sealed c PoW))
+
+-- | A data fetcher is a wrapper around an action that given a (singleton)
+-- 'SProtocolRequest' and a set of arguments, returns a 'ProtocolResponse'
+-- while producting some side-effect in @m@.
+newtype DataFetcher c tx s m =
+    DataFetcher {
+      fetch :: forall r. ActivePeer c
+            -> SProtocolRequest r
+            -> ProtocolRequestArgs r
+            -> m (Either Timeout (ProtocolResponse c tx s r))
+                }
+
+-- | An opaque reference to a record of functions which can be used to query
+-- the network and the other parts of the system for information.
+data SyncContext c tx s m = SyncContext
+    { scNu               :: Nu
```

Comment:
I'd prefer if this was just a `Natural` - we already know it's `nu` based on the name, but now the type tells me nothing because I don't know what `Nu` could be..

Reply:
Agreed.